crates/sqlexec/src/remote/broadcast/exchange_exec.rs

```diff
@@ -0,0 +1,499 @@
+                    "Reading first batch error: {e}"
+                )))
+            }
+            None => {
+                return Err(ExecError::RemoteSession(
+                    "Missing first batch on input stream".to_string(),
+                ))
+            }
+        };
+
+        Ok(ClientExchangeRecvStream {
+            session_id,
+            broadcast_id,
+            stream: input,
+            buf: batches,
+            schema,
+        })
+    }
+
+    pub fn session_id(&self) -> Uuid {
+        self.session_id
+    }
+
+    pub fn broadcast_id(&self) -> Uuid {
+        self.broadcast_id
+    }
+
+    fn read_arrow_ipc(
+        buf: Vec<u8>,
+    ) -> DataFusionResult<impl Iterator<Item = DataFusionResult<RecordBatch>>> {
+        let cursor = Cursor::new(buf);
+        let reader = IpcFileReader::try_new(cursor, None)?;
+        Ok(reader.into_iter().map(|result| match result {
+            Ok(batch) => Ok(batch),
+            Err(e) => Err(DataFusionError::ArrowError(e)),
+        }))
+    }
+}
+
+// TODO: StreamReader instead of FileReader.
+impl Stream for ClientExchangeRecvStream {
+    type Item = DataFusionResult<RecordBatch>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        // Check buffer.
+        if let Some(batch) = self.buf.pop_front() {
+            return Poll::Ready(Some(batch));
+        }
+
+        // Pull from stream.
+        //
+        // Trigger oneshot on error or fi the stream completes.
+        match self.stream.poll_next_unpin(cx) {
+            Poll::Ready(Some(req)) => match req {
+                Ok(req) => {
+                    let batches = match Self::read_arrow_ipc(req.arrow_ipc) {
+                        Ok(iter) => iter,
+                        Err(e) => {
+                            return Poll::Ready(Some(Err(e)));
+                        }
+                    };
+
+                    // Extend out buffer with batches from the ipc reader.
+                    self.buf.extend(batches);
+
+                    // See if we got anything.
+                    match self.buf.pop_front() {
+                        Some(result) => Poll::Ready(Some(result)),
+                        None => Poll::Ready(None),
+                    }
+                }
+                Err(e) => Poll::Ready(Some(Err(DataFusionError::Execution(format!(
+                    "failed to pull next batch from stream: {e}"
+                ))))),
+            },
+            Poll::Pending => Poll::Pending,
+            Poll::Ready(None) => Poll::Ready(None),
+        }
+    }
+}
+
+impl RecordBatchStream for ClientExchangeRecvStream {
+    fn schema(&self) -> Arc<Schema> {
+        self.schema.clone()
+    }
+}
+
+/// The actual execution plan for reading batches from the client.
+#[derive(Debug)]
+pub struct ClientExchangeRecvExec {
+    schema: Arc<Schema>,
+    stream: Mutex<Option<ClientExchangeRecvStream>>,
```

Comment:
won't this make it impossible to serialize the physical plan?

Reply:
Sort of. I'm currently approaching this where the logical plans I'm defining for this are the "serializable" physical plans right now (https://github.com/GlareDB/glaredb/pull/1559/files#diff-9fb769d3f9c2de52b94a5d19ad4b3a25c53fcf5aed31ef06c5eb44e163b0e3c6R85-R88). When we have the proper extensions ready, we could make the appropriate changes to this where get the stream from the task context which would remove the need for this.

Reply:
E.g. we would make an extension that would get us this: https://github.com/GlareDB/glaredb/pull/1559/files#diff-e3453d481fa630bb58c1e5fe35d4730223f55457bdb744b654f42b60a9639541R30-R33

Reply:
Actually I'll go ahead and make that change right now to see how well it works.